utils/s2n_asn1_time.c

```diff
@@ -0,0 +1,255 @@
+/*
+ * Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "utils/s2n_asn1_time.h"
+#include "s2n_safety.h"
+
+#include <time.h>
+#include <ctype.h>
+
+typedef enum parser_state {
+    ON_YEAR_DIGIT_1 = 0,
+    ON_YEAR_DIGIT_2,
+    ON_YEAR_DIGIT_3,
+    ON_YEAR_DIGIT_4,
+    ON_MONTH_DIGIT_1,
+    ON_MONTH_DIGIT_2,
+    ON_DAY_DIGIT_1,
+    ON_DAY_DIGIT_2,
+    ON_HOUR_DIGIT_1,
+    ON_HOUR_DIGIT_2,
+    ON_MINUTE_DIGIT_1,
+    ON_MINUTE_DIGIT_2,
+    ON_SECOND_DIGIT_1,
+    ON_SECOND_DIGIT_2,
+    ON_SUBSECOND,
+    ON_TIMEZONE,
+    ON_OFFSET_HOURS_DIGIT_1,
+    ON_OFFSET_HOURS_DIGIT_2,
+    ON_OFFSET_MINUTES_DIGIT_1,
+    ON_OFFSET_MINUTES_DIGIT_2,
+    FINISHED,
+    PARSE_ERROR
+} parser_state;
+
+static inline long get_gmt_offset(struct tm *time) {
+
+#if defined(__USE_BSD)
+    return time->tm_gmtoff;
+#else
+    return time->__tm_gmtoff;
+#endif
+}
+
+static inline long get_current_timezone_offset(void) {
+    struct tm time_ptr;
+    time_t raw_time;
+    time(&raw_time);
+    localtime_r(&raw_time, &time_ptr);
+    return get_gmt_offset(&time_ptr);
+}
+
+int s2n_asn1_time_to_nano_since_epoch_ticks(const char *asn1_time, uint32_t len, uint64_t *ticks) {
+
+    //figure out if we are on something than UTC since timegm is not supported everywhere.
+    long gmt_offset_current = get_current_timezone_offset();
+
+    uint32_t str_len = len;
+    parser_state state = ON_YEAR_DIGIT_1;
+
+    struct tm time = {.tm_hour = 0, .tm_isdst = -1, .tm_mday = 0, .tm_min = 0, .tm_mon = 0,
+            .tm_sec = 0, .tm_wday = 0, .tm_yday = 0, .tm_year = 0,
+    };
+
+    uint32_t current_pos = 0;
+    uint8_t offset_negative = 0;
+    uint8_t local_time_assumed = 1;
+    uint8_t current_digit = 0;
+    long offset_hours = 0;
+    long offset_minutes = 0;
+
+    //this is just a standard state machine for ASN1 date format... nothing special.
+    //just do a character at a time and change the state per character encountered.
+    //when finished the above time structure should be filled in along with some
+    //crazy timezone info we'll need shortly afterwards.
+    while (state < FINISHED && current_pos < str_len) {
+        char current_char = asn1_time[current_pos];
+        switch (state) {
+            case ON_YEAR_DIGIT_1:
```

Comment:
Unless OpenSSL is dong some automatic conversion for you, this will only handle an ASN.1 `GeneralizedTime` String with 4 digits for the year, but not a `UTCTime` String with 2 digits for the year, I think we need to handle both.

From the book [Implementing SSL/TLS, Chapter 5, page 233](https://doc.lagout.org/security/Crypto/Implementing%20SSL_TLS%20Using%20Cryptography%20and%20PKI.pdf):
> So, with a two-digit year, the client has to do a bit of detective work to figure out if "35" expired a very, very long time ago, or if it will expire in 25 years. Because no X.509 certificates were issued in 1935, it’s safe to assume that a year of "35" means 2035. In fact, the specification mandates that all certificates issued before 2050 must use UTCTime, so if the year is less than "50", it’s in the 21st century. After the year 2050, CA’s are supposed to begin using GeneralizedTime, with a four-digit year.

Links:
 - https://www.obj-sys.com/asn1tutorial/node14.html
 - https://www.obj-sys.com/asn1tutorial/node15.html

Reply:
https://www.openssl.org/docs/man1.1.0/crypto/OCSP_single_get0_status.html

The libcrypto X.509 code only exposes GeneralizedTime. Considering that, do you think we should still add support for UTC time?